include/itkAttenuationImageFilter.hxx

```diff
@@ -0,0 +1,316 @@
+  }
+
+  const MaskImageType *  maskImage = this->GetMaskImage();
+  const InputImageType * input = this->GetInput();
+  OutputImageType *      output = this->GetOutput();
+
+  ImageLinearConstIteratorWithIndex<TInputImage> it(input, regionForThread);
+  it.SetDirection(m_Direction);
+  it.GoToBegin();
+
+  unsigned int   inclusionLength;
+  InputIndexType start, end;
+
+  const float scanStepMM = input->GetSpacing()[m_Direction];
+
+  // do the work
+  while (!it.IsAtEnd())
+  {
+    inclusionLength = 0;
+    while (!it.IsAtEndOfLine())
+    {
+      // Advance until an inclusion is found
+      InputIndexType index = it.GetIndex();
+      if (ThreadedIsIncluded(index) && inclusionLength == 0)
+      {
+        // Step into inclusion
+        start = it.GetIndex();
+      }
+      else if (!ThreadedIsIncluded(index) && inclusionLength > 0)
+      {
+        // Stay at last pixel in the inclusion
+        end = it.GetIndex();
+        end[m_Direction] -= 1;
+
+        // Adjust for pixel padding
+        start[m_Direction] += m_PadLowerBounds;
+        end[m_Direction] -= m_PadUpperBounds;
+
+        // Estimate attenuation for each inclusion pixel
+        // with respect to the last pixel in the inclusion
+        while (start[m_Direction] < end[m_Direction])
+        {
+          // If no fixed estimation depth is set or too few pixels remain for fixed-depth estimation
+          // then take the attenuation between the given pixel and the last pixel in the inclusion
+          InputIndexType target = end;
+
+          if (m_FixedEstimationDepth != 0 && end[m_Direction] - start[m_Direction] > m_FixedEstimationDepth)
+          {
+            target[m_Direction] = start[m_Direction] + m_FixedEstimationDepth;
+          }
+
+          float estimatedAttenuation = ComputeAttenuation(target, start);
+
+          // Update the corresponding pixel in the metric image
+          if (estimatedAttenuation > 0.0 || m_ConsiderNegativeAttenuations)
+          {
+            // Assignment is thread-safe for internal operations because
+            // each thread writes only within its allotted output region.
+            output->SetPixel(start, estimatedAttenuation);
+          }
+
+          ++start[m_Direction];
+        }
+
+        inclusionLength = 0;
+      }
+
+      if (ThreadedIsIncluded(index))
+      {
+        ++inclusionLength;
+      }
+
+      ++it;
+    }
+    it.NextLine();
+  }
+};
+
+template <typename TInputImage, typename TOutputImage, typename TMaskImage>
+typename AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::OutputPixelType
+AttenuationImageFilter<TInputImage, TOutputImage, TMaskImage>::ComputeAttenuation(const InputIndexType & end,
+                                                                                  const InputIndexType & start) const
+{
+  const float nyquistFrequency = m_SamplingFrequencyMHz / 2;
+
+  // Number and width of RF spectra frequency bins over the range (0, nyquist_frequency]
```

Comment:
Should the range be inclusive on both ends? I.E. `[0, nyquist_frequency]`. The rest looks good.

Reply:
No, good question, `itkSpectraImageFilter.h` specifically drops DC components so that content at `0MHz` is excluded from spectra vector images. Please see [`itkSpectraImageFilter.h`](https://github.com/KitwareMedical/ITKUltrasound/blob/master/include/itkSpectra1DImageFilter.h#L48).